Custom Contracts: MDD SQL Error

Mission Control has to manually add `EncounterLayer` to the MDD since ModTek, nor the HBS loader handles that.

I need to improve it as it can sometimes cause a SQL error on BT boot as a race condition or some kind of clash.

Investigate and fix.

```
Uploading Crash Report
Unity [ERROR] SqliteException: SQLite error
not an error
 
(Filename: C:\buildslave\unity\build\Runtime/Export/Debug.bindings.h Line: 45)

SqliteException: SQLite error
not an error
  at Mono.Data.Sqlite.SQLite3.Reset (Mono.Data.Sqlite.SqliteStatement stmt) [0x00096] in <99f7d8f9a65b4d91ae5b1d55c424866f>:0 
  at Mono.Data.Sqlite.SQLite3.Step (Mono.Data.Sqlite.SqliteStatement stmt) [0x00046] in <99f7d8f9a65b4d91ae5b1d55c424866f>:0 
  at Mono.Data.Sqlite.SqliteDataReader.NextResult () [0x00129] in <99f7d8f9a65b4d91ae5b1d55c424866f>:0 
  at (wrapper remoting-invoke-with-check) Mono.Data.Sqlite.SqliteDataReader.NextResult()
  at Mono.Data.Sqlite.SqliteDataReader..ctor (Mono.Data.Sqlite.SqliteCommand cmd, System.Data.CommandBehavior behave) [0x00051] in <99f7d8f9a65b4d91ae5b1d55c424866f>:0 
  at (wrapper remoting-invoke-with-check) Mono.Data.Sqlite.SqliteDataReader..ctor(Mono.Data.Sqlite.SqliteCommand,System.Data.CommandBehavior)
  at Mono.Data.Sqlite.SqliteCommand.ExecuteReader (System.Data.CommandBehavior behavior) [0x00006] in <99f7d8f9a65b4d91ae5b1d55c424866f>:0 
  at Mono.Data.Sqlite.SqliteCommand.ExecuteNonQuery () [0x00000] in <99f7d8f9a65b4d91ae5b1d55c424866f>:0
at BattleTech.Data.MetadataDatabase.DetachMDDFromInMemoryDB () [0x00017] in <4184af8dbeb44635831353f4d349631c>:0 
  at BattleTech.Data.MetadataDatabase.WriteInMemoryDBToDisk (System.String filePath) [0x000fa] in <4184af8dbeb44635831353f4d349631c>:0 
  at MissionControl.DataManager.LoadCustomContractTypes () [0x00072] in <f15ecd11837d4fde8e934f3fce3abebb>:0 
  at MissionControl.DataManager.LoadDeferredDefs () [0x00016] in <f15ecd11837d4fde8e934f3fce3abebb>:0 
  at MissionControl.Patches.MainMenuInitPatch.Postfix (BattleTech.UI.MainMenu __instance) [0x00071] in <f15ecd11837d4fde8e934f3fce3abebb>:0 
  at (wrapper dynamic-method) BattleTech.UI.MainMenu.Init_Patch2(object)
  at BattleTech.UI.UIManager.AddModule (BattleTech.UI.UIModule module, BattleTech.UI.UIManager+UINode node, System.Int32 depth, System.Boolean resort) [0x0008f] in <4184af8dbeb44635831353f4d349631c>:0 
  at BattleTech.UI.UIManager.GetOrCreateAtNode (System.Type moduleType, System.String prefabName, BattleTech.UI.UIManager+UINode node, System.Boolean tryGet, System.Boolean resort) [0x00067] in <4184af8dbeb44635831353f4d349631c>:0 
  at BattleTech.UI.UIManager.CreateUIModule (System.Type moduleType, System.String prefabOverride, System.Boolean resort) [0x0000a] in <4184af8dbeb44635831353f4d349631c>:0 
  at BattleTech.UI.ModuleAutoLauncher.Launch () [0x00056] in <4184af8dbeb44635831353f4d349631c>:0 
  at BattleTech.UI.ModuleAutoLauncher+<Start>d__7.MoveNext () [0x00046] in <4184af8dbeb44635831353f4d349631c>:0 
  at UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) [0x00027] in <9a6ceb7f013a4a599806f492fc789483>:0 
```

--- a/src/Core/DataManager.cs
+++ b/src/Core/DataManager.cs
@@ -1,7 +1,10 @@
+using UnityEngine;
+
 using System;
 using System.IO;
-using System.Collections.Generic;
 using System.Globalization;
+using System.Collections;
+using System.Collections.Generic;
 
 using Newtonsoft.Json;
 using Newtonsoft.Json.Linq;
@@ -102,6 +105,11 @@ namespace MissionControl {
     }
 
     private void LoadCustomContractTypes() {
+      UnityGameInstance.Instance.StartCoroutine(WriteMDDToDisk());
+    }
+
+    IEnumerator WriteMDDToDisk() {
+      yield return new WaitForSeconds(2);
       MetadataDatabase mdd = MetadataDatabase.Instance;
       List<ContractType_MDD> contractTypes = mdd.GetCustomContractTypes();
       Main.LogDebug($"[DataManager.LoadCustomContractTypes] Loading '{contractTypes.Count}' custom contract type(s)");

--- a/src/Core/DataManager.cs
+++ b/src/Core/DataManager.cs
@@ -2,6 +2,7 @@ using UnityEngine;
 
 using System;
 using System.IO;
+using System.Threading;
 using System.Globalization;
 using System.Collections;
 using System.Collections.Generic;
@@ -105,14 +106,21 @@ namespace MissionControl {
     }
 
     private void LoadCustomContractTypes() {
-      UnityGameInstance.Instance.StartCoroutine(WriteMDDToDisk());
+      new Thread(new ThreadStart(this.WriteMDDToDisk)).Start();
+      UnityGameInstance.Instance.StartCoroutine(ReportOnLoadedCustomContractType());
     }
 
-    IEnumerator WriteMDDToDisk() {
-      yield return new WaitForSeconds(2);
+    IEnumerator ReportOnLoadedCustomContractType() {
+      yield return new WaitForSeconds(5); // A largish amount of time to ensure contract types have been saved by then
       MetadataDatabase mdd = MetadataDatabase.Instance;
       List<ContractType_MDD> contractTypes = mdd.GetCustomContractTypes();
-      Main.LogDebug($"[DataManager.LoadCustomContractTypes] Loading '{contractTypes.Count}' custom contract type(s)");
+      Main.LogDebug($"[DataManager.LoadCustomContractTypes] Loaded '{contractTypes.Count}' custom contract type(s)");
+    }
+
+    private void WriteMDDToDisk() {
+      Thread.Sleep(Main.Settings.ContractTypeLoaderWait);
+      MetadataDatabase mdd = MetadataDatabase.Instance;
+      List<ContractType_MDD> contractTypes = mdd.GetCustomContractTypes();
       foreach (ContractType_MDD contractType in contractTypes) {
         AddContractType(mdd, contractType);
         LoadEncounterLayers(contractType.Name);

--- a/src/Core/Settings/Settings.cs
+++ b/src/Core/Settings/Settings.cs
@@ -10,6 +10,9 @@ namespace MissionControl.Config {
     [JsonProperty("Debug")]
     public DebugSettings Debug { get; set; } = new DebugSettings();
 
+    [JsonProperty("ContractTypeLoaderWait")]
+    public int ContractTypeLoaderWait { get; set; } = 1000;
+
     [JsonProperty("VersionCheck")]
     public bool VersionCheck { get; set; } = true;
 

